### PR TITLE
define and implement cloud InstanceV2

### PR DESCRIPTION
**Pull request by gongguan**
/kind feature

Ref #90652

Part of #91319

```release-note
None
```

### CODE REVIEW

File: staging/src/k8s.io/cloud-provider/cloud.go [hunk 1]
```diff
@@ -49,6 +49,8 @@ type Interface interface {
 	LoadBalancer() (LoadBalancer, bool)
 	// Instances returns an instances interface. Also returns true if the interface is supported, false otherwise.
 	Instances() (Instances, bool)
+	// InstancesV2 is an external-only implementation for instances. Also returns true if the interface is supported, false otherwise.
```
Comment (wojtek-t) [inline staging/src/k8s.io/cloud-provider/cloud.go:52]: What do you mean by "external-only" ?

File: staging/src/k8s.io/cloud-provider/cloud.go [hunk 2]
```diff
@@ -186,6 +188,18 @@ type Instances interface {
 	InstanceShutdownByProviderID(ctx context.Context, providerID string) (bool, error)
 }
 
+// InstancesV2 is an abstract, pluggable interface for sets of instances.
+// Unlike Instances, it is a external-only implementation which should not be called the components like
```
Comment (wojtek-t) [inline staging/src/k8s.io/cloud-provider/cloud.go:192]: I think I don't understand this sentence.

Comment (wojtek-t) [reply staging/src/k8s.io/cloud-provider/cloud.go:192]: @andrewsykim - to help figuring out the best wording here

File: staging/src/k8s.io/cloud-provider/cloud.go (hunk 1 above)
Comment (gongguan) [reply staging/src/k8s.io/cloud-provider/cloud.go:52]: Might it be internal-only? I'd like to ask for @andrewsykim 's original thought.

Comment (andrewsykim) [reply staging/src/k8s.io/cloud-provider/cloud.go:52]: Sorry, the comment in the example I shared was not fully thought out. 

"external-only" meant it should only be called by cloud-controller-manager at the moment. Cloud node controller in https://github.com/kubernetes/kubernetes/pull/91319 only runs in cloud-controller-manager and not in kube-controller-manager or kubelet.

Comment (gongguan) [reply staging/src/k8s.io/cloud-provider/cloud.go]: No sorry, your example helps me a lot to understand your thought. Updated the comments, PTAL

**Comment by wojtek-t** [comment]
/approve
/retest

@andrewsykim - does that look ok to you too?

File: staging/src/k8s.io/legacy-cloud-providers/aws/aws.go [hunk 3]
```diff
@@ -1394,6 +1394,11 @@ func (c *Cloud) Instances() (cloudprovider.Instances, bool) {
 	return c, true
 }
 
+// InstancesV2 returns an implementation of InstancesV2 for Amazon Web Services.
+func (c *Cloud) InstancesV2() (cloudprovider.InstancesV2, bool) {
+	return c, true
```
Comment (andrewsykim) [inline staging/src/k8s.io/legacy-cloud-providers/aws/aws.go:1399]: set this to false and let aws OWNERs flip this on?

File: staging/src/k8s.io/legacy-cloud-providers/azure/azure.go [hunk 4]
```diff
@@ -659,6 +659,11 @@ func (az *Cloud) Instances() (cloudprovider.Instances, bool) {
 	return az, true
 }
 
+// InstancesV2 returns an instancesV2 interface. Also returns true if the interface is supported, false otherwise.
+func (az *Cloud) InstancesV2() (cloudprovider.InstancesV2, bool) {
+	return az, true
```
Comment (andrewsykim) [inline staging/src/k8s.io/legacy-cloud-providers/azure/azure.go:664]: set this to false and let Azure OWNERs flip this on?

File: staging/src/k8s.io/legacy-cloud-providers/gce/gce.go [hunk 5]
```diff
@@ -659,6 +659,11 @@ func (g *Cloud) Instances() (cloudprovider.Instances, bool) {
 	return g, true
 }
 
+// InstancesV2 returns an implementation of InstancesV2 for Google Compute Engine.
+func (g *Cloud) InstancesV2() (cloudprovider.InstancesV2, bool) {
+	return g, true
```
Comment (andrewsykim) [inline staging/src/k8s.io/legacy-cloud-providers/gce/gce.go:664]: set this to false and let GCE OWNERs flip this on?

File: staging/src/k8s.io/legacy-cloud-providers/openstack/openstack_instances.go [hunk 6]
```diff
@@ -62,6 +61,16 @@ func (os *OpenStack) Instances() (cloudprovider.Instances, bool) {
...
 }
 
+// Instances returns an implementation of Instances for OpenStack.
+func (os *OpenStack) Instances() (cloudprovider.Instances, bool) {
+	return os.instances()
+}
+
+// InstancesV2 returns an implementation of InstancesV2 for OpenStack.
+func (os *OpenStack) InstancesV2() (cloudprovider.InstancesV2, bool) {
+	return os.instances()
```
Comment (andrewsykim) [inline staging/src/k8s.io/legacy-cloud-providers/openstack/openstack_instances.go:71]: Ideally this shouldn't call the same thing as Instances

File: staging/src/k8s.io/legacy-cloud-providers/vsphere/vsphere.go [hunk 7]
```diff
@@ -560,6 +560,11 @@ func (vs *VSphere) Instances() (cloudprovider.Instances, bool) {
 	return vs, true
 }
 
+// InstancesV2 returns an implementation of InstancesV2 for vSphere.
+func (vs *VSphere) InstancesV2() (cloudprovider.InstancesV2, bool) {
+	return vs, true
```
Comment (andrewsykim) [inline staging/src/k8s.io/legacy-cloud-providers/vsphere/vsphere.go:565]: set this to `false` and let Azure OWNERs flip this on?

**Review by andrewsykim — COMMENTED** [review]
Overall lgtm, it might be better to leave InstancesV2 off for the in-tree providers and let the individual providers turn it on.

File: staging/src/k8s.io/legacy-cloud-providers/aws/aws.go (hunk 3 above)
Comment (gongguan) [reply staging/src/k8s.io/legacy-cloud-providers/aws/aws.go]: All set as false, it's a method with low risk.

File: staging/src/k8s.io/legacy-cloud-providers/gce/gce.go (hunk 5 above)
Comment (gongguan) [reply staging/src/k8s.io/legacy-cloud-providers/gce/gce.go]: Set as false, while I thought @wojtek-t might want gce use InstancesV2() to reduce api calls, right?

Comment (wojtek-t) [reply staging/src/k8s.io/legacy-cloud-providers/gce/gce.go]: We can do that when we will be migrating the controller. So I'm fine with leaving false for now.

**Review by andrewsykim — COMMENTED** [review]
/approve
/lgtm

Thanks for iterating on this @gongguan!

File: staging/src/k8s.io/legacy-cloud-providers/vsphere/vsphere.go (hunk 7 above)
Comment (feiskyer) [reply staging/src/k8s.io/legacy-cloud-providers/vsphere/vsphere.go]: nice for this, would add InstancesV2 support for Azure.

Comment (wojtek-t) [reply staging/src/k8s.io/legacy-cloud-providers/vsphere/vsphere.go]: It's already added - all built-in providers have this already implemented. It's just a matter of changing this value here.

**Pull request merged**